{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "f0545de3",
   "metadata": {},
   "source": [
    "# Generate F plasmid for GSMM"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "456453ee",
   "metadata": {},
   "source": [
    "### Load packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9336e51c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cobra\n",
    "from cobra import Model, Reaction, Metabolite\n",
    "import numpy as np \n",
    "import pandas as pd\n",
    "import re\n",
    "import string\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "from cobra.io import load_json_model, save_json_model, load_matlab_model, save_matlab_model, read_sbml_model, write_sbml_model "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b285300",
   "metadata": {},
   "source": [
    "### Make metabolite dictionaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eab29cd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "metDict = {\n",
    "    'atp': 'datp_c',        # ATP,              ChEBI 15422\n",
    "    'ctp': 'dctp_c',        # CTP,              ChEBI 17677\n",
    "    'gtp': 'dgtp_c',        # GTP,              ChEBI 15996\n",
    "    'ttp': 'dttp_c',        # UTP,              ChEBI 15713\n",
    "    'A': 'ala__L_c',        # Alaline,          ChEBI 16977\n",
    "    'R': 'arg__L_c',        # Arginine,         ChEBI 16467\n",
    "    'N': 'asn__L_c',        # Asparagine,       ChEBI 17196\n",
    "    'D': 'asp__L_c',        # Aspartate,        ChEBI 17053\n",
    "    'C': 'cys__L_c',        # Cysteine,         ChEBI 17561\n",
    "    'Q': 'gln__L_c',        # Glutamine,        ChEBI 18050\n",
    "    'E': 'glu__L_c',        # Glutamate,        ChEBI 16015\n",
    "    'G': 'gly_c',          # Glycine,          ChEBI 15428\n",
    "    'H': 'his__L_c',        # Histidine,        ChEBI 15971\n",
    "    'I': 'ile__L_c',        # Isoleucine,       ChEBI 17191\n",
    "    'L': 'leu__L_c',        # Leucine,          ChEBI 15603\n",
    "    'K': 'lys__L_c',        # Lysine,           ChEBI 18019\n",
    "    'M': 'met__L_c',        # Methionine,       ChEBI 16643\n",
    "    'F': 'phe__L_c',        # Phenylalanine,    ChEBI 17295\n",
    "    'P': 'pro__L_c',        # Proline,          ChEBI 17203\n",
    "    'S': 'ser__L_c',        # Serine,           ChEBI 17115\n",
    "    'T': 'thr__L_c',        # Threonine,        ChEBI 16857\n",
    "    'W': 'trp__L_c',        # Tryptophan,       ChEBI 16828\n",
    "    'Y': 'tyr__L_c',        # Tyrosine,         ChEBI 17895\n",
    "    'V': 'val__L_c',        # Valine,           ChEBI 16414\n",
    "    'h2o': 'h2o_c',        # H2O\n",
    "    'adp': 'adp_c',        # ADP\n",
    "    'Pi': 'pi_c',          # Phosphate\n",
    "    'h': 'h_c',            # Hydrogen [Proton]\n",
    "    'PPi': 'ppi_c',        # Pyrophosphate\n",
    "}\n",
    "# Nucleotide dictionary with molecular weights\n",
    "# Source: ChEBI https://www.ebi.ac.uk/chebi/\n",
    "ntpsDict = {\n",
    "    'atp': 507.181,     # ATP,              ChEBI 15422\n",
    "    'gtp': 483.15644,   # GTP,              ChEBI 17677\n",
    "    'ctp': 523.18062,   # CTP,              ChEBI 15996\n",
    "    'ttp': 498.16770,   # TTP,              ChEBI 15713\n",
    "}\n",
    "# Amino Acids dictionary with molecular weights\n",
    "# Source: ChEBI https://www.ebi.ac.uk/chebi/\n",
    "aaDict = {\n",
    "    'A': 89.09322,      # Alanine,          ChEBI 16977\n",
    "    'R': 174.201,       # Arginine,         ChEBI 16467\n",
    "    'N': 132.118,       # Asparagine,       ChEBI 17196\n",
    "    'D': 133.1027,      # Aspartate,        ChEBI 17053\n",
    "    'C': 121.158,       # Cysteine,         ChEBI 17561\n",
    "    'Q': 146.14458,     # Glutamine,        ChEBI 18050\n",
    "    'E': 147.1293,      # Glutamate,        ChEBI 16015\n",
    "    'G': 75.06664,      # Glycine,          ChEBI 15428\n",
    "    'H': 155.15468,     # Histidine,        ChEBI 15971\n",
    "    'I': 131.17296,     # Isoleucine,       ChEBI 17191\n",
    "    'L': 131.17296,     # Leucine,          ChEBI 15603\n",
    "    'K': 146.18764,     # Lysine,           ChEBI 18019\n",
    "    'M': 149.21238,     # Methionine,       ChEBI 16643\n",
    "    'F': 165.18918,     # Phenylalanine,    ChEBI 17295\n",
    "    'P': 115.1305,      # Proline,          ChEBI 17203\n",
    "    'S': 105.09262,     # Serine,           ChEBI 17115\n",
    "    'T': 119.1192,      # Threonine,        ChEBI 16857\n",
    "    'W': 204.22526,     # Tryptophan,       ChEBI 16828\n",
    "    'Y': 181.18858,     # Tyrosine,         ChEBI 17895\n",
    "    'V': 117.14638,     # Valine,           ChEBI 16414\n",
    "}\n",
    "# Misc. dictionary with molecular weights\n",
    "# Source: ChEBI https://www.ebi.ac.uk/chebi/\n",
    "miscDict = {\n",
    "    'PPi': 173.94332,   # Pyrophosphate,    ChEBI 18361\n",
    "}\n",
    "# Nucleotides List\n",
    "ntpsMets    = list(ntpsDict.keys())\n",
    "# Amino Acids List\n",
    "aaMets      = list(aaDict.keys())\n",
    "# Avogadro's Number\n",
    "N_A         = 6.0221409e+23\n",
    "# Energy requirement coefficients\n",
    "# Source: Haynie, D. (2008). Biological Thermodynamics (2nd ed.). Cambridge: Cambridge University Press.\n",
    "k_atp       = 4\n",
    "k_ppi       = 1\n",
    "\n",
    "datp_c = Metabolite(metDict['atp'])\n",
    "dctp_c = Metabolite(metDict['ctp'])\n",
    "dgtp_c = Metabolite(metDict['gtp'])\n",
    "dttp_c = Metabolite(metDict['ttp'])\n",
    "ala__L_c = Metabolite(metDict['A'])\n",
    "arg__L_c = Metabolite(metDict['R'])\n",
    "asn__L_c = Metabolite(metDict['N'])\n",
    "asp__L_c = Metabolite(metDict['D'])\n",
    "cys__L_c = Metabolite(metDict['C'])\n",
    "gln__L_c = Metabolite(metDict['Q'])\n",
    "glu__L_c = Metabolite(metDict['E'])\n",
    "gly_c = Metabolite(metDict['G'])\n",
    "his__L_c = Metabolite(metDict['H'])\n",
    "ile__L_c = Metabolite(metDict['I'])\n",
    "leu__L_c = Metabolite(metDict['L'])\n",
    "lys__L_c = Metabolite(metDict['K'])\n",
    "met__L_c = Metabolite(metDict['M'])\n",
    "phe__L_c = Metabolite(metDict['F'])\n",
    "pro__L_c = Metabolite(metDict['P'])\n",
    "ser__L_c = Metabolite(metDict['S'])\n",
    "thr__L_c = Metabolite(metDict['T'])\n",
    "trp__L_c = Metabolite(metDict['W'])\n",
    "tyr__L_c = Metabolite(metDict['Y'])\n",
    "val__L_c = Metabolite(metDict['V'])\n",
    "adp_c = Metabolite(metDict['adp'])\n",
    "h2o_c = Metabolite(metDict['h2o'])\n",
    "h_c   = Metabolite(metDict['h'])\n",
    "pi_c  = Metabolite(metDict['Pi'])\n",
    "ppi_c = Metabolite(metDict['PPi'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6dcc417",
   "metadata": {},
   "source": [
    "### Make pilus gene dictionary "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d74685e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#based on Braggagnolo et al 2020\n",
    "max_pilin_units = 1e6\n",
    "single_pilus_dict = {\n",
    "    'FinO' : max_pilin_units * 0.0001, #transcriptional regulation\n",
    "    'TraJ' : max_pilin_units * 0.0001, #transcriptional regulation\n",
    "    'TraR' : max_pilin_units * 0.0001, #transcriptional regulation\n",
    "    'TraA' : max_pilin_units, #propilin component, 12.8 units per helical turn in a 5 start helix, \n",
    "    #65 units per 13.2 Angstroms of length, maximum pilus length of 20microns leads to 1e6 max per pilus \n",
    "    #from Braggagnolo et al 2020\n",
    "    'TraQ' : max_pilin_units * 0.25, #transient attachment to traA so can be reused\n",
    "    'TraX' : max_pilin_units * 0.5, #acetylates propilin\n",
    "    'TraB' : max_pilin_units * 0.001, #t4ss core protein\n",
    "    'TraK' : max_pilin_units * 0.001, #t4ss core protein\n",
    "    'TraV' : max_pilin_units * 0.001, #t4ss core protein\n",
    "    'TraP' : max_pilin_units * 0.01, #assembly/extension\n",
    "    'TraE' : max_pilin_units * 0.01, #assembly/extension\n",
    "    'TraL' : max_pilin_units * 0.01, #assembly/extension\n",
    "    'TraC' : max_pilin_units * 0.01, #assembly/extension\n",
    "    'TraW' : max_pilin_units * 0.01, #assembly/extension\n",
    "    'TrbC' : max_pilin_units * 0.01, #assembly/extension\n",
    "    'TraF' : max_pilin_units * 0.01, #assembly/extension\n",
    "    'TrbB' : max_pilin_units * 0.01, #assembly/extension\n",
    "    'TraH' : max_pilin_units * 0.01, #retraction\n",
    "    'TrbI' : max_pilin_units * 0.01, #retraction\n",
    "    'TraN' : max_pilin_units * 0.001, #mating pair stabilization\n",
    "    'TraG' : max_pilin_units * 0.001, #mating pair stabilization\n",
    "    'TraU' : max_pilin_units * 0.001, #mating pair stabilization\n",
    "    'TraI' : max_pilin_units * 0.001, #relaxase\n",
    "    'TraY' : max_pilin_units * 0.001, #relaxase\n",
    "    'TraM' : max_pilin_units * 0.001, #relaxase\n",
    "    'TraD' : max_pilin_units * 0.001, #relaxase\n",
    "    'TraT' : max_pilin_units * 0.0001, #surface exclusion\n",
    "    'TraS' : max_pilin_units * 0.0001, #entry exclusion\n",
    "    'TrbD' : max_pilin_units * 0.00001, #unknown function\n",
    "    'TrbE' : max_pilin_units * 0.00001, #unknown function\n",
    "    'TrbF' : max_pilin_units * 0.00001, #unknown function\n",
    "    'TrbA' : max_pilin_units * 0.00001, #unknown function\n",
    "    'TrbG' : max_pilin_units * 0.00001, #unknown function\n",
    "    'TrbJ' : max_pilin_units * 0.00001 #unknown function\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0a99dca-104f-415b-bcdf-3556bc660fc7",
   "metadata": {},
   "source": [
    "### Make M13 gene dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "16b9e442-9b82-47a1-83f2-4968cc9b0977",
   "metadata": {},
   "outputs": [],
   "source": [
    "#based on Wang et al\n",
    "phage_dict = {\n",
    "    'III' : 5, #minor capsid protein\n",
    "    'VI' : 5, #minor capsid protein\n",
    "    'VII' : 5, #minor capsid protein\n",
    "    'IX' : 5, #minor capsid protein\n",
    "    'VIII' : 2700, #major capsin protein\n",
    "    'II' : 10, #initiates DNA replication, koskoe\n",
    "    'X' : 10, #inhibits replication\n",
    "    'V' : 1500, #packaging protein, encapsulates DNA\n",
    "    'I' : 15, #part of gated channel\n",
    "    'IV' : 15 #conners et al, gated channel \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c289e70e",
   "metadata": {},
   "source": [
    "### Define the plasmid and phage generating functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "65780948",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_model_with_plasmid(number_of_pili, model):\n",
    "    #Import the F plasmid and Tn10 sequences from Genbank\n",
    "    plasmid = \"../models/genbank/f-plasmid.txt\"\n",
    "    with open(plasmid,newline = '') as pf:\n",
    "        plasmidFile = pf.readlines()\n",
    "    plasmidData = str(plasmidFile)\n",
    "\n",
    "    tn10 = \"../models/genbank/tn10.txt\"\n",
    "    with open(tn10,newline = '') as tf:\n",
    "        tn10File = tf.readlines()\n",
    "    tn10Data = str(tn10File)\n",
    "    \n",
    "    #Calculate the nucleotide code of the F plasmid and Tn10 \n",
    "    startG  = [jj for jj, s in enumerate(plasmidFile) if 'ORIGIN' in s]\n",
    "    endG    = [jj for jj, s in enumerate(plasmidFile) if '//' in s]\n",
    "    startG  = int(''.join(map(str,startG)))\n",
    "    endG    = int(''.join(map(str,endG)))\n",
    "    startG  = startG + 1\n",
    "\n",
    "    startG_t  = [jj for jj, s in enumerate(tn10File) if 'ORIGIN' in s]\n",
    "    endG_t    = [jj for jj, s in enumerate(tn10File) if '//' in s]\n",
    "    startG_t  = int(''.join(map(str,startG_t)))\n",
    "    endG_t    = int(''.join(map(str,endG_t)))\n",
    "    startG_t  = startG_t + 1\n",
    "    \n",
    "    regex           = re.compile('[^a-zA-Z]')\n",
    "    plasmidGenome     = str(''.join(plasmidFile[startG:endG]))\n",
    "    plasmidGenome     = regex.sub('',plasmidGenome)\n",
    "\n",
    "    tn10Genome     = str(''.join(tn10File[startG_t:endG_t]))\n",
    "    tn10Genome     = regex.sub('',tn10Genome)\n",
    "    \n",
    "    Cg = 1 #genome copy number, for F plasmid usually only 1 and never more than 2\n",
    "    countA  = plasmidGenome.count('a')\n",
    "    countC  = plasmidGenome.count('c')\n",
    "    countG  = plasmidGenome.count('g')\n",
    "    countT  = plasmidGenome.count('t')   \n",
    "    antiA   = countT\n",
    "    antiC   = countG\n",
    "    antiG   = countC\n",
    "    antiT   = countA\n",
    "\n",
    "    countA_t  = tn10Genome.count('a')\n",
    "    countC_t   = tn10Genome.count('c')\n",
    "    countG_t   = tn10Genome.count('g')\n",
    "    countT_t   = tn10Genome.count('t')   \n",
    "    antiA_t    = countT_t \n",
    "    antiC_t    = countG_t \n",
    "    antiG_t    = countC_t \n",
    "    antiT_t    = countA_t \n",
    "\n",
    "    totNTPS   = (Cg * (countA + countC + countG + countT + antiA + antiC + antiG + antiT + countA_t + countC_t + countG_t + countT_t + antiA_t + antiC_t + antiG_t + antiT_t))\n",
    "\n",
    "    # mol.ntps/mol.plasmid\n",
    "    V_a = (Cg*(countA + antiA + countA_t + antiA_t))\n",
    "    V_c = (Cg*(countC + antiC + countC_t + antiC_t))\n",
    "    V_g = (Cg*(countG + antiG + countG_t + antiG_t))\n",
    "    V_t = (Cg*(countT + antiT + countT_t + antiT_t))\n",
    "    # g.ntps/mol.plasmid\n",
    "    G_a = V_a * ntpsDict[\"atp\"]\n",
    "    G_c = V_c * ntpsDict[\"ctp\"]\n",
    "    G_g = V_g * ntpsDict[\"gtp\"]\n",
    "    G_t = V_t * ntpsDict[\"ttp\"]\n",
    "    \n",
    "    # Calculate the protein cost of the pilus-producing regions of the F plasmid\n",
    "    pilus_genes = list(single_pilus_dict.keys())\n",
    "    pilus_genes.remove('FinO')\n",
    "    pilus_genes.remove('TraX')\n",
    "    pilus_genes.remove('TraR')\n",
    "    pilus_gene_IDs = dict.fromkeys(pilus_genes)\n",
    "\n",
    "    for g in range(0,len(pilus_genes)):\n",
    "        proteinString = '/product=\"conjugal transfer protein ' + pilus_genes[g] + '\"'\n",
    "        tempLine = [jj for jj, s in enumerate(plasmidFile) if proteinString in s]\n",
    "        val = tempLine[0]\n",
    "        pilus_gene_IDs[pilus_genes[g]] = val\n",
    "\n",
    "    pilus_gene_IDs[\"TraX\"] = [jj for jj, s in enumerate(plasmidFile) if \"WP_000205715.1\" in s][0]\n",
    "    \n",
    "    linesOfInterest = list(pilus_gene_IDs.values())\n",
    "    linesOfInterest.sort()\n",
    "    conjugalProteins = plasmidFile[linesOfInterest[0]:linesOfInterest[-1] + 25]\n",
    "\n",
    "    translationLines = [jj for jj, s in enumerate(conjugalProteins) if \"/translation\" in s]\n",
    "    geneLines = [jj for jj, s in enumerate(conjugalProteins) if \"gene            \" in s] #number of tab spaces matters!\n",
    "    productLines = [jj for jj, s in enumerate(conjugalProteins) if '/product=\"conjugal' in s]\n",
    "\n",
    "    translationDictionary = {}\n",
    "    for i in range(0, len(productLines) - 1):\n",
    "        start = productLines[i]\n",
    "        end = productLines[i + 1]\n",
    "        temp = conjugalProteins[start:end]\n",
    "        translationStart = [jj for jj, s in enumerate(temp) if \"/translation\" in s][0]\n",
    "        translationEnd = [jj for jj, s in enumerate(temp) if \"gene            \" in s][0]\n",
    "        productString = [jj for jj, s in enumerate(temp) if '/product=\"conjugal transfer protein' in s][0]\n",
    "        geneString = temp[productString].split()[-1]\n",
    "        geneString = geneString.translate(str.maketrans('', '', string.punctuation))\n",
    "        translationRegion = temp[translationStart:translationEnd]\n",
    "        npReg = re.compile('/translation=')\n",
    "        tempgene  = str(translationRegion)\n",
    "        tempgene  = npReg.sub('',tempgene)\n",
    "        tempgene  = regex.sub('',tempgene)\n",
    "        translationDictionary[geneString] = tempgene\n",
    "\n",
    "    endDict = 31\n",
    "    start = productLines[endDict]\n",
    "    end = productLines[endDict] + 10\n",
    "    temp = conjugalProteins[start:end]\n",
    "    translationStart = [jj for jj, s in enumerate(temp) if \"/translation\" in s][0]\n",
    "    translationEnd = [jj for jj, s in enumerate(temp) if \"gene            \" in s][0]\n",
    "    geneString = \"TraX\"\n",
    "    translationRegion = temp[translationStart:translationEnd]\n",
    "    npReg = re.compile('/translation=')\n",
    "    tempgene  = str(translationRegion)\n",
    "    tempgene  = npReg.sub('',tempgene)\n",
    "    tempgene  = regex.sub('',tempgene)\n",
    "    translationDictionary[geneString] = tempgene\n",
    "    \n",
    "    pilus_proteins = list(translationDictionary.keys())\n",
    "    np_list = np.zeros((20,1))\n",
    "    for xx in pilus_proteins:\n",
    "        aaCount = np.zeros((20,1))\n",
    "        protein = translationDictionary[xx]\n",
    "        for ii in range(len(aaMets)):\n",
    "            aaCount[ii,0] = protein.count(aaMets[ii])\n",
    "        finalarray = aaCount * single_pilus_dict[xx]\n",
    "        numpy_adjusted = np.array(finalarray)\n",
    "        np_list = np_list + finalarray\n",
    "\n",
    "    pilus_protein_content = np_list * number_of_pili\n",
    "    \n",
    "    # Calculate the protein cost of the non pilus-producing regions of the F plasmid\n",
    "    singleProductLines = [jj for jj, s in enumerate(plasmidFile) if '/product=' in s]\n",
    "    singleTranslationLines = [jj for jj, s in enumerate(plasmidFile) if \"/translation\" in s]\n",
    "    singleGeneLines = [jj for jj, s in enumerate(plasmidFile) if \"gene            \" in s] #number of tab spaces matters!\n",
    "\n",
    "    fullTranslationString = ''\n",
    "    for i in range(0, len(singleProductLines) - 1):\n",
    "        start = singleProductLines[i]\n",
    "        end = singleProductLines[i + 1]\n",
    "        temp = plasmidFile[start:end]\n",
    "        if not [jj for jj, s in enumerate(temp) if \"/translation\" in s]:\n",
    "            continue\n",
    "        else:\n",
    "            translationStart = [jj for jj, s in enumerate(temp) if \"/translation\" in s][0]\n",
    "            translationEnd = [jj for jj, s in enumerate(temp) if \"gene            \" in s][0]\n",
    "            translationRegion = temp[translationStart:translationEnd]\n",
    "            npReg = re.compile('/translation=')\n",
    "            tempgene  = str(translationRegion)\n",
    "            tempgene  = npReg.sub('',tempgene)\n",
    "            tempgene  = regex.sub('',tempgene)\n",
    "            fullTranslationString = fullTranslationString + tempgene\n",
    "    \n",
    "    endDict = 241\n",
    "    start = singleProductLines[endDict]\n",
    "    end = singleProductLines[endDict] + 10\n",
    "    temp = plasmidFile[start:end]\n",
    "    translationStart = [jj for jj, s in enumerate(temp) if \"/translation\" in s][0]\n",
    "    translationEnd = [jj for jj, s in enumerate(temp) if \"ORIGIN\" in s][0]\n",
    "    translationRegion = temp[translationStart:translationEnd]\n",
    "    npReg = re.compile('/translation=')\n",
    "    tempgene  = str(translationRegion)\n",
    "    tempgene  = npReg.sub('',tempgene)\n",
    "    tempgene  = regex.sub('',tempgene)\n",
    "    fullTranslationString = fullTranslationString + tempgene\n",
    "\n",
    "\n",
    "    nonPilusCount = np.zeros((20,1))\n",
    "    for ii in range(len(aaMets)):\n",
    "        nonPilusCount[ii,0] = fullTranslationString.count(aaMets[ii])\n",
    "\n",
    "    tn10ProductLines = [jj for jj, s in enumerate(tn10File) if '/product=' in s]\n",
    "    tn10TranslationLines = [jj for jj, s in enumerate(tn10File) if \"/translation\" in s]\n",
    "    tn10GeneLines = [jj for jj, s in enumerate(tn10File) if \"gene            \" in s] #number of tab spaces matters!\n",
    "\n",
    "    fulltn10TranslationString = ''\n",
    "    for i in range(0, len(tn10ProductLines) - 1):\n",
    "        start = tn10ProductLines[i]\n",
    "        end = tn10ProductLines[i + 1]\n",
    "        temp = tn10File[start:end]\n",
    "        if not [jj for jj, s in enumerate(temp) if \"/translation\" in s]:\n",
    "            continue\n",
    "        else:\n",
    "            translationStart = [jj for jj, s in enumerate(temp) if \"/translation\" in s][0]\n",
    "            translationEnd = [jj for jj, s in enumerate(temp) if \"gene            \" in s][0]\n",
    "            translationRegion = temp[translationStart:translationEnd]\n",
    "            npReg = re.compile('/translation=')\n",
    "            tempgene  = str(translationRegion)\n",
    "            tempgene  = npReg.sub('',tempgene)\n",
    "            tempgene  = regex.sub('',tempgene)\n",
    "            fulltn10TranslationString = fulltn10TranslationString + tempgene\n",
    "            \n",
    "    endDict = 4\n",
    "    start = tn10ProductLines[endDict]\n",
    "    end = tn10ProductLines[endDict] + 35\n",
    "    temp = tn10File[start:end]\n",
    "    translationStart = [jj for jj, s in enumerate(temp) if \"/translation\" in s][0]\n",
    "    translationEnd = [jj for jj, s in enumerate(temp) if \"mobile_element\" in s][0]\n",
    "    translationRegion = temp[translationStart:translationEnd]\n",
    "    npReg = re.compile('/translation=')\n",
    "    tempgene  = str(translationRegion)\n",
    "    tempgene  = npReg.sub('',tempgene)\n",
    "    tempgene  = regex.sub('',tempgene)\n",
    "    fulltn10TranslationString = fulltn10TranslationString + tempgene\n",
    "\n",
    "    tn10Count = np.zeros((20,1))\n",
    "    for ii in range(len(aaMets)):\n",
    "        tn10Count[ii,0] = fulltn10TranslationString.count(aaMets[ii])\n",
    "    \n",
    "    # Calculate total amino acid cost\n",
    "    totAA = tn10Count + pilus_protein_content + nonPilusCount\n",
    "    \n",
    "    # mol.aa/mol.plasmid\n",
    "    V_aa    = np.zeros((20,1))\n",
    "    for ii in range(len(aaMets)):\n",
    "        V_aa[ii,0] = totAA[ii]\n",
    "    # g.a/mol.plasmid\n",
    "    G_aa    = np.zeros((20,1))\n",
    "    for ii in range(len(aaMets)):\n",
    "        G_aa[ii,0] = V_aa[ii] * aaDict[aaMets[ii]]\n",
    "        \n",
    "    # Combine the final counts for nucleotide and protein costs\n",
    "    \n",
    "    M_v = (G_a + G_c + G_g + G_t) + G_aa.sum()\n",
    "    \n",
    "    # Nucleotides [mmol.ntps/g.virus]\n",
    "    S_atp = 1000 * (V_a/M_v)\n",
    "    S_ctp = 1000 * (V_c/M_v)\n",
    "    S_gtp = 1000 * (V_g/M_v)\n",
    "    S_ttp = 1000 * (V_t/M_v)\n",
    "\n",
    "    # Amino acids [mmol.aa/g.virus]\n",
    "    S_aa    = np.zeros((20,1))\n",
    "    for ii in range(len(aaMets)):\n",
    "        S_aa[ii] = 1000 * (V_aa[ii]/M_v)\n",
    "\n",
    "    # Genome: Phosphodiester bond formation products [Pyrophosphate]\n",
    "    genTemp = (((countA + countC + countG + countT + countA_t + countC_t + countG_t + countT_t) * k_ppi) - k_ppi)\n",
    "    genRep  = (((antiA + antiC + antiG + antiT + antiA_t + antiC_t + antiG_t + antiT_t) * k_ppi) - k_ppi)\n",
    "    genTot  = genTemp + genRep\n",
    "    V_ppi   = genTot\n",
    "    S_ppi   = 1000 * (V_ppi/M_v)\n",
    "\n",
    "    # Protome: Peptide bond formation [ATP + H2O]\n",
    "    # Note: ATP used in this process is denoated as ATPe/Ae [e = energy version]\n",
    "    ppAe    = ((pilus_protein_content.sum() * k_atp) - k_atp)\n",
    "    tnAe    = ((tn10Count.sum() * k_atp) - k_atp)\n",
    "    npAe    = ((nonPilusCount.sum() * k_atp) - k_atp)\n",
    "    ppTot   = ppAe + tnAe + npAe\n",
    "    V_Ae    = ppTot\n",
    "    S_Ae    = 1000 * (V_Ae/M_v)\n",
    "    \n",
    "    # Generate reaction\n",
    "    \n",
    "    # Left-hand terms: Nucleotides\n",
    "    # Note: ATP term is a summation of genome and energy requirements\n",
    "    S_ATP   = (S_atp + S_Ae) * -1\n",
    "    S_CTP   = S_ctp * -1\n",
    "    S_GTP   = S_gtp * -1\n",
    "    S_TTP   = S_ttp * -1\n",
    "    # Left-hand terms: Amino Acids\n",
    "    S_AA    = S_aa * -1\n",
    "    S_AAf   = dict()\n",
    "    for ii in range(len(aaMets)):\n",
    "        S_AAf[aaMets[ii]] = S_AA[ii,0]\n",
    "    # Left-hand terms: Energy Requirements\n",
    "    S_H2O   = S_Ae * -1\n",
    "    # Right-hand terms: Energy Requirements\n",
    "    S_ADP   = S_Ae\n",
    "    S_Pi    = S_Ae\n",
    "    S_H     = S_Ae\n",
    "    S_PPi   = S_ppi\n",
    "    \n",
    "    Name = \"plasmid\"\n",
    "    reaction_name       = Name + '_F'\n",
    "    plasmid_reaction      = Reaction(reaction_name)\n",
    "    plasmid_reaction.lower_bound              = 0.0\n",
    "    plasmid_reaction.upper_bound              = 1000.0\n",
    "    plasmid_reaction.add_metabolites(({\n",
    "        datp_c: S_ATP,\n",
    "        dctp_c: S_CTP,\n",
    "        dgtp_c: S_GTP,\n",
    "        dttp_c: S_TTP,\n",
    "        ala__L_c: S_AAf['A'],\n",
    "        arg__L_c: S_AAf['R'],\n",
    "        asn__L_c: S_AAf['N'],\n",
    "        asp__L_c: S_AAf['D'],\n",
    "        cys__L_c: S_AAf['C'],\n",
    "        gln__L_c: S_AAf['Q'],\n",
    "        glu__L_c: S_AAf['E'],\n",
    "        gly_c: S_AAf['G'],\n",
    "        his__L_c: S_AAf['H'],\n",
    "        ile__L_c: S_AAf['I'],\n",
    "        leu__L_c: S_AAf['L'],\n",
    "        lys__L_c: S_AAf['K'],\n",
    "        met__L_c: S_AAf['M'],\n",
    "        phe__L_c: S_AAf['F'],\n",
    "        pro__L_c: S_AAf['P'],\n",
    "        ser__L_c: S_AAf['S'],\n",
    "        thr__L_c: S_AAf['T'],\n",
    "        trp__L_c: S_AAf['W'],\n",
    "        tyr__L_c: S_AAf['Y'],\n",
    "        val__L_c: S_AAf['V'],\n",
    "        h2o_c: S_H2O,\n",
    "        adp_c: S_ADP,\n",
    "        pi_c:  S_Pi,\n",
    "        h_c:   S_H,\n",
    "        ppi_c: S_PPi}))\n",
    "    \n",
    "    hostModel  = cobra.io.read_sbml_model(model)\n",
    "    \n",
    "    hostModel.add_reactions([plasmid_reaction])\n",
    "    hostModel.reactions[-1].reversibility = False\n",
    "    \n",
    "    return(hostModel)\n",
    "\n",
    "def generate_model_with_phage(model, progeny_number = 1):\n",
    "    phage = \"../models/genbank/m13.gbff\"\n",
    "    with open(phage,newline = '') as pf:\n",
    "        phageFile = pf.readlines()\n",
    "    phageData = str(phageFile)\n",
    "    \n",
    "    startG  = [jj for jj, s in enumerate(phageFile) if 'ORIGIN' in s]\n",
    "    endG    = [jj for jj, s in enumerate(phageFile) if '//' in s]\n",
    "    startG  = int(''.join(map(str,startG)))\n",
    "    endG    = int(''.join(map(str,endG)))\n",
    "    startG  = startG + 1\n",
    "    \n",
    "    regex           = re.compile('[^a-zA-Z]')\n",
    "    phageGenome     = str(''.join(phageFile[startG:endG]))\n",
    "    phageGenome     = regex.sub('',phageGenome)\n",
    "    \n",
    "    Cg = 1 #genome copy number per phage particle\n",
    "    countA  = phageGenome.count('a')\n",
    "    countC  = phageGenome.count('c')\n",
    "    countG  = phageGenome.count('g')\n",
    "    countT  = phageGenome.count('t')   \n",
    "    antiA   = countT\n",
    "    antiC   = countG\n",
    "    antiG   = countC\n",
    "    antiT   = countA\n",
    "    \n",
    "    totNTPS   = (Cg * (countA + countC + countG + countT + antiA + antiC + antiG + antiT))\n",
    "    \n",
    "    # mol.ntps/mol.plasmid\n",
    "    V_a = (Cg*(countA + antiA))\n",
    "    V_c = (Cg*(countC + antiC))\n",
    "    V_g = (Cg*(countG + antiG))\n",
    "    V_t = (Cg*(countT + antiT))\n",
    "    # g.ntps/mol.plasmid\n",
    "    G_a = V_a * ntpsDict[\"atp\"]\n",
    "    G_c = V_c * ntpsDict[\"ctp\"]\n",
    "    G_g = V_g * ntpsDict[\"gtp\"]\n",
    "    G_t = V_t * ntpsDict[\"ttp\"]\n",
    "    \n",
    "    phage_genes = list(phage_dict.keys())\n",
    "    phage_gene_IDs = dict.fromkeys(phage_genes)\n",
    "    \n",
    "    for g in range(0,len(phage_genes)):\n",
    "        proteinString = '/gene=\"' + phage_genes[g] + '\"'\n",
    "        tempLine = [jj for jj, s in enumerate(phageFile) if proteinString in s]\n",
    "        val = tempLine[0]\n",
    "        phage_gene_IDs[phage_genes[g]] = val\n",
    "    \n",
    "    linesOfInterest = list(phage_gene_IDs.values())\n",
    "    linesOfInterest.sort()\n",
    "    Proteins = phageFile[linesOfInterest[0]-1:linesOfInterest[-1] + 16]\n",
    "    \n",
    "    translationLines = [jj for jj, s in enumerate(Proteins) if \"/translation\" in s]\n",
    "    geneLines = [jj for jj, s in enumerate(Proteins) if \"gene            \" in s] #number of tab spaces matters!\n",
    "    productLines = [jj + 1 for jj, s in enumerate(Proteins) if \"gene            \" in s]\n",
    "    \n",
    "    productLines.append(len(Proteins))\n",
    "    \n",
    "    translationDictionary = {}\n",
    "    for i in range(0, len(productLines) - 1):\n",
    "        start = productLines[i]\n",
    "        end = productLines[i + 1]\n",
    "        temp = Proteins[start:end]\n",
    "        translationEnd = len(temp) - 1\n",
    "        translationStart = [jj for jj, s in enumerate(temp) if \"/translation\" in s][0]\n",
    "        productString = [jj for jj, s in enumerate(temp) if '/gene=\"' in s][0]\n",
    "        geneString = temp[productString].split()[-1]\n",
    "        noPunctString = geneString.translate(str.maketrans('', '', string.punctuation))\n",
    "        noGeneString = noPunctString.replace(\"gene\", \"\")\n",
    "        translationRegion = temp[translationStart:translationEnd]\n",
    "        npReg = re.compile('/translation=')\n",
    "        tempgene  = str(translationRegion)\n",
    "        tempgene  = npReg.sub('',tempgene)\n",
    "        tempgene  = regex.sub('',tempgene)\n",
    "        translationDictionary[noGeneString] = tempgene\n",
    "    \n",
    "    phage_proteins = list(translationDictionary.keys())\n",
    "    np_list = np.zeros((20,1))\n",
    "    for xx in phage_proteins:\n",
    "        aaCount = np.zeros((20,1))\n",
    "        protein = translationDictionary[xx]\n",
    "        for ii in range(len(aaMets)):\n",
    "            aaCount[ii,0] = protein.count(aaMets[ii])\n",
    "        finalarray = aaCount * phage_dict[xx]\n",
    "        numpy_adjusted = np.array(finalarray)\n",
    "        np_list = np_list + finalarray\n",
    "\n",
    "    total_protein_content = np_list * progeny_number\n",
    "    \n",
    "    # mol.aa/mol.plasmid\n",
    "    V_aa    = np.zeros((20,1))\n",
    "    for ii in range(len(aaMets)):\n",
    "        V_aa[ii,0] = total_protein_content[ii]\n",
    "    # g.a/mol.plasmid\n",
    "    G_aa    = np.zeros((20,1))\n",
    "    for ii in range(len(aaMets)):\n",
    "        G_aa[ii,0] = V_aa[ii] * aaDict[aaMets[ii]]\n",
    "        \n",
    "    # Combine the final counts for nucleotide and protein costs\n",
    "    M_v = (G_a + G_c + G_g + G_t) + G_aa.sum()\n",
    "    \n",
    "    # Nucleotides [mmol.ntps/g.virus]\n",
    "    S_atp = 1000 * (V_a/M_v)\n",
    "    S_ctp = 1000 * (V_c/M_v)\n",
    "    S_gtp = 1000 * (V_g/M_v)\n",
    "    S_ttp = 1000 * (V_t/M_v)\n",
    "    \n",
    "    # Amino acids [mmol.aa/g.virus]\n",
    "    S_aa    = np.zeros((20,1))\n",
    "    for ii in range(len(aaMets)):\n",
    "        S_aa[ii] = 1000 * (V_aa[ii]/M_v)\n",
    "    \n",
    "    # Genome: Phosphodiester bond formation products [Pyrophosphate]\n",
    "    genTemp = (((countA + countC + countG + countT) * k_ppi) - k_ppi)\n",
    "    genRep  = (((antiA + antiC + antiG + antiT) * k_ppi) - k_ppi)\n",
    "    genTot  = genTemp + genRep\n",
    "    V_ppi   = genTot\n",
    "    S_ppi   = 1000 * (V_ppi/M_v)\n",
    "    \n",
    "    # Protome: Peptide bond formation [ATP + H2O]\n",
    "    # Note: ATP used in this process is denoated as ATPe/Ae [e = energy version]\n",
    "    ppAe    = ((total_protein_content.sum() * k_atp) - k_atp)\n",
    "    ppTot   = ppAe\n",
    "    V_Ae    = ppTot\n",
    "    S_Ae    = 1000 * (V_Ae/M_v)\n",
    "    \n",
    "    # Generate reaction\n",
    "    \n",
    "    # Left-hand terms: Nucleotides\n",
    "    # Note: ATP term is a summation of genome and energy requirements\n",
    "    S_ATP   = (S_atp + S_Ae) * -1\n",
    "    S_CTP   = S_ctp * -1\n",
    "    S_GTP   = S_gtp * -1\n",
    "    S_TTP   = S_ttp * -1\n",
    "    # Left-hand terms: Amino Acids\n",
    "    S_AA    = S_aa * -1\n",
    "    S_AAf   = dict()\n",
    "    for ii in range(len(aaMets)):\n",
    "        S_AAf[aaMets[ii]] = S_AA[ii,0]\n",
    "    # Left-hand terms: Energy Requirements\n",
    "    S_H2O   = S_Ae * -1\n",
    "    # Right-hand terms: Energy Requirements\n",
    "    S_ADP   = S_Ae\n",
    "    S_Pi    = S_Ae\n",
    "    S_H     = S_Ae\n",
    "    S_PPi   = S_ppi\n",
    "    \n",
    "    Name = \"phage\"\n",
    "    reaction_name       = Name + '_M13'\n",
    "    phage_reaction      = Reaction(reaction_name)\n",
    "    phage_reaction.lower_bound              = 0.0\n",
    "    phage_reaction.upper_bound              = 1000.0\n",
    "    phage_reaction.add_metabolites(({\n",
    "        datp_c: S_ATP,\n",
    "        dctp_c: S_CTP,\n",
    "        dgtp_c: S_GTP,\n",
    "        dttp_c: S_TTP,\n",
    "        ala__L_c: S_AAf['A'],\n",
    "        arg__L_c: S_AAf['R'],\n",
    "        asn__L_c: S_AAf['N'],\n",
    "        asp__L_c: S_AAf['D'],\n",
    "        cys__L_c: S_AAf['C'],\n",
    "        gln__L_c: S_AAf['Q'],\n",
    "        glu__L_c: S_AAf['E'],\n",
    "        gly_c: S_AAf['G'],\n",
    "        his__L_c: S_AAf['H'],\n",
    "        ile__L_c: S_AAf['I'],\n",
    "        leu__L_c: S_AAf['L'],\n",
    "        lys__L_c: S_AAf['K'],\n",
    "        met__L_c: S_AAf['M'],\n",
    "        phe__L_c: S_AAf['F'],\n",
    "        pro__L_c: S_AAf['P'],\n",
    "        ser__L_c: S_AAf['S'],\n",
    "        thr__L_c: S_AAf['T'],\n",
    "        trp__L_c: S_AAf['W'],\n",
    "        tyr__L_c: S_AAf['Y'],\n",
    "        val__L_c: S_AAf['V'],\n",
    "        h2o_c: S_H2O,\n",
    "        adp_c: S_ADP,\n",
    "        pi_c:  S_Pi,\n",
    "        h_c:   S_H,\n",
    "        ppi_c: S_PPi}))\n",
    "    \n",
    "    hostModel  = cobra.io.read_sbml_model(model)\n",
    "    \n",
    "    hostModel.add_reactions([phage_reaction])\n",
    "    hostModel.reactions[-1].reversibility = False\n",
    "    return(hostModel)\n",
    "\n",
    "def get_host_optimization(model, host_reaction):\n",
    "    HVM = model\n",
    "    HostRxn = host_reaction\n",
    "    try:\n",
    "        intTest = int(HostRxn)\n",
    "        hostIdx = HostRxn\n",
    "    except:\n",
    "        for ii in range(len(HVM.reactions)):\n",
    "            if HostRxn in str(HVM.reactions[ii]):\n",
    "                hostIdx = ii\n",
    "    HVM.objective = host_reaction\n",
    "    hostSol     = HVM.optimize()\n",
    "    hostX       = hostSol.fluxes                 # Host-optimal flux vector\n",
    "    hostF       = hostSol.objective_value \n",
    "    return(hostF)\n",
    "\n",
    "def get_plasmid_optimization(model, plasmid_reaction):\n",
    "    HVM = model\n",
    "    plasmidIdx    = len(HVM.reactions) - 1 \n",
    "    HVM.reactions[plasmidIdx].objective_coefficient    = 0.0\n",
    "    # Host optimisation\n",
    "    HVM.objective = plasmid_reaction\n",
    "    # Record Optima\n",
    "    plasmidSol     = HVM.optimize()\n",
    "    plasmidX       = plasmidSol.fluxes                 # Host-optimal flux vector\n",
    "    plasmidF       = plasmidSol.objective_value \n",
    "    return(plasmidF)\n",
    "\n",
    "def change_plasmid_bound(model, lowerbound, biomass_reaction):\n",
    "    HVM = model\n",
    "    plasmidIdx    = len(HVM.reactions) - 1    \n",
    "    # Host optimisation\n",
    "    HVM.objective = biomass_reaction\n",
    "    # Record Optima\n",
    "    HVM.reactions[plasmidIdx].lower_bound = lowerbound\n",
    "    plasmidSol     = HVM.optimize()\n",
    "    plasmidX       = plasmidSol.fluxes                 # Host-optimal flux vector\n",
    "    plasmidF       = plasmidSol.objective_value \n",
    "    return(plasmidF)\n",
    "\n",
    "def get_lowerbound_data(model, plasmid_reaction, plasmid_step_size, biomass_reaction):\n",
    "    maximum_plasmid = get_plasmid_optimization(model, plasmid_reaction)\n",
    "    values = {}\n",
    "    for i in np.arange(0, maximum_plasmid + plasmid_step_size, plasmid_step_size):\n",
    "        data = change_plasmid_bound(model, i, biomass_reaction)\n",
    "        values[i] = data\n",
    "        if i > 0 and list(values.values())[-1] == list(values.values())[-2]:\n",
    "            print(\"Biomass has stopped changing\")\n",
    "            break\n",
    "    return(values)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87cc6ead-8067-40ca-b8d0-6cbd767b0120",
   "metadata": {},
   "source": [
    "### Generate and validate model with phage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "fd0d8ca5-cce0-4ad5-bda0-fdbf54bfb056",
   "metadata": {},
   "outputs": [],
   "source": [
    "phageInfectedHost = generate_model_with_phage(\"../models/iJO1366.xml\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3235fc26",
   "metadata": {},
   "source": [
    "### Generate and validate model with F plasmid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "2b7ab050",
   "metadata": {},
   "outputs": [],
   "source": [
    "pili_values = [0, 2, 10, 500, 1e6]\n",
    "change_bound_and_pili_count = pd.DataFrame()\n",
    "for i in pili_values:\n",
    "    plasmidInfectedHost = generate_model_with_plasmid(i, \"../models/iJO1366.xml\")\n",
    "    temp = get_lowerbound_data(plasmidInfectedHost, \"plasmid_F\", 0.1, \"BIOMASS_Ec_iJO1366_core_53p95M\")\n",
    "    temp_df = pd.DataFrame(temp.items())\n",
    "    temp_df[\"pili_count\"] = i\n",
    "    change_bound_and_pili_count = pd.concat([change_bound_and_pili_count, temp_df])\n",
    "\n",
    "change_bound_and_pili_count = change_bound_and_pili_count.rename(columns={0: 'lower_plasmid_bound', 1: 'cell_biomass'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "720b5dc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lower_plasmid_bound</th>\n",
       "      <th>cell_biomass</th>\n",
       "      <th>pili_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.1</td>\n",
       "      <td>0.944080</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.1</td>\n",
       "      <td>0.871570</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.1</td>\n",
       "      <td>0.871383</td>\n",
       "      <td>10.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.1</td>\n",
       "      <td>0.871337</td>\n",
       "      <td>500.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.1</td>\n",
       "      <td>0.871336</td>\n",
       "      <td>1000000.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   lower_plasmid_bound  cell_biomass  pili_count\n",
       "1                  0.1      0.944080         0.0\n",
       "1                  0.1      0.871570         2.0\n",
       "1                  0.1      0.871383        10.0\n",
       "1                  0.1      0.871337       500.0\n",
       "1                  0.1      0.871336   1000000.0"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "change_bound_and_pili_count[change_bound_and_pili_count[\"lower_plasmid_bound\"] == 0.1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9da3ce93",
   "metadata": {},
   "source": [
    "### Export the model with the F plasmid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "fc78815a",
   "metadata": {},
   "outputs": [],
   "source": [
    "plasmidInfectedHost = generate_model_with_plasmid(10, \"../models/Ecoli_F1_V3.xml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "09f37923",
   "metadata": {},
   "outputs": [],
   "source": [
    "write_sbml_model(plasmidInfectedHost, \"../models/iJO1366_plasmid.xml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "2b47dd2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "phageInfectedHost = generate_model_with_phage(\"../models/iJO1366_plasmid.xml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "83d85ead-b7a2-49b1-9bfb-e905718567b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "write_sbml_model(phageInfectedHost, \"../models/iJO1366_plasmid_phage.xml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5911472-f649-4f33-b73d-2f3624cd78b7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
